### PR TITLE
fix(pipeline): forward termination-aware discount to HordeActorCriticAgent in horde_ac mode

### LINKED ISSUE #2344: AlbertaPipeline control_mode="horde_ac" drops `terminated`: the value head bootstraps and the actor trace survives every episode boundary
`AlbertaPipeline.update` documents a `terminated` argument ("Scalar termination
flag (`0.0` or `1.0`)") and validates it as a float32 scalar restricted to
`{0, 1}`. On the default `control_mode="sarsa"` path it is forwarded into
`step4_update` → `SARSAAgent.update`, where it skips bootstrapping
(`skip_bootstrap = (safe_terminated != 0) | (gamma_arr == 0.0)`,
`alberta_framework/core/sarsa.py:945`) and zeroes the eligibility traces
(`sarsa.py:1005-1006`).

On the registered `control_mode="horde_ac"` path it is dropped on the floor.
`alberta_framework/pipeline.py:1576-1581` calls

```python
ac_result = ac.update(
    ac_state,
    reward,
    features,
    auxiliary_cumulants=auxiliary_cumulants,
)
```

with no `discount`, so `HordeActorCriticAgent.update` falls back to
`value_discount = self._critic.horde_spec.gammas[cfg.value_head_index]`
(`alberta_framework/core/horde_actor_critic.py:1001-1003`) on *every* step,
terminal or not.

Two things are consequently wrong at a terminal transition:

1. **The value head bootstraps through termination.** The critic's TD target
   for the value demon is `r + gamma * V(s')` instead of `r`.
2. **The actor eligibility trace survives the boundary.** In
   `HordeActorCriticAgent.update`, `actor_decay = value_discount * cfg.actor_lamda`
   and `carry_traces = value_discount != 0.0` (`horde_actor_critic.py:1044,1061`),
   so with `discount = 0` the actor trace is zeroed — but the pipeline never
   supplies it, and the trace is carried into the next episode instead.

The mechanism to fix it already exists and was built for exactly this caller.
`HordeActorCriticAgent.update` takes `discount: Array | None` — "Optional
scalar per-transition value-head discount" — and routes to
`HordeLearner.update_with_discounts`, whose docstring says it "lets control
adapters zero the value head at episode boundaries while keeping the Horde's
fixed GVF metadata and per-head trace decay intact"
(`alberta_framework/core/horde.py:554-557`). The pipeline is such a control
adapter and is the only in-tree one that does not use it.

## Reproduction

`alberta_framework/pipeline.py` at `c9aba7b5`, identity Step 2 features,
`horde.gammas=(0.9, 0.5)`, `value_head_index=0`, `actor_lamda=0.8`:

```python
state = pipeline.init(jr.key(0), jnp.asarray([0.2, -0.1, 0.4], dtype=jnp.float32))
obs = jnp.asarray([0.1, 0.3, -0.2], dtype=jnp.float32)
reward = jnp.asarray(0.5, dtype=jnp.float32)
cumulants = jnp.asarray([0.5, -0.2], dtype=jnp.float32)
for term in (0.0, 1.0):
    res = pipeline.update(state, obs, reward, jnp.asarray(term, dtype=jnp.float32), cumulants)
    print(term, res.horde_td_targets, res.state.control_state.actor_trace_weights)
```

```
0.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
1.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
```

The two lines are byte-identical: `terminated` has **no effect whatsoever** on
this control path. The value-head target should be `0.5` (the reward alone) in
the terminal case; it is `0.5 + 0.9 * V(s') = 0.5 + 0.9 * (-0.1718) = 0.3454`.
The actor trace should be empty; it is not.

## Why no test caught it

Every existing pipeline test passes `terminated = 0.0`
(`tests/test_pipeline.py:237`, `:280`, `:324`, `:412`, `:484`, `:540`, `:580`;
the array-run helper builds `terminated = jnp.zeros(n_steps)` at `:632`). `terminated = 1.0` is exercised nowhere in
the pipeline suite, on either control mode, so the entire termination semantics
of the documented flag are untested and the `horde_ac` branch's untested path
is silently wrong.

## Scope

`AlbertaPipeline.run_arrays` delegates to `AlbertaPipeline.update`, so array
runs and `run_pipeline_smoke` inherit the same defect. Any measurement taken
through the pipeline with `control_mode="horde_ac"` on an episodic stream
learns terminal transitions as if the episode continued. Auxiliary GVF demons
are *not* affected and must keep their own configured gammas — only the
value head's discount is a per-transition control quantity.

## Proposed fix

Pass the terminal-aware discount in the `horde_ac` branch:

```python
value_gamma = self._horde.horde_spec.gammas[value_index]
control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))
ac_result = ac.update(..., discount=control_discount)
```

At `terminated == 0` this is a bit-exact reduction: `update_with_discounts`
computes `bootstrap = jnp.where(discounts == 0.0, 0.0, discounts * next_preds)`
from the same `gammas` vector the `discount is None` branch uses, and the extra
finite/range head-validity conditions are satisfied by the configured gammas.

Lane/metric: this is a measurement-trustworthiness fix, not a benchmark climb —
no `outputs/` artifact changes and no registered evidence source is touched.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2344

## Summary
In `alberta_framework/pipeline.py`:
On the `control_mode="horde_ac"` path, `AlbertaPipeline.update` called `HordeActorCriticAgent.update` without passing `discount`. As a result, the agent fell back to the value demon's default gamma on every transition, causing the value head to bootstrap through episode termination and actor eligibility traces to survive episode boundaries.

## Proposed Changes
1. Derived `control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))` from `self._horde.horde_spec.gammas[value_index]`.
2. Forwarded `discount=control_discount` to `HordeActorCriticAgent.update`.
3. Added unit tests in `tests/test_pipeline.py` verifying that terminal transitions (`terminated=1.0`) receive unbootstrapped TD targets ($r$) and zero actor traces across the boundary.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_pipeline.py` (184/184 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/pipeline.py tests/test_pipeline.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/pipeline.py` (clean).

### CODE REVIEW

**Review by lumix5 — COMMENTED** [review]
**Disposition: close as duplicate of #2528** — same termination-aware `discount` wiring for the `horde_ac` control path; #2528 predates it by ~19 hours, carries the 184-test isolated verification (terminal TD target = bare reward, actor traces zeroed) and approval. This copy's larger diff is otherwise formatting noise.

**Review by lumix5 — APPROVED** [review]
Independent verification at the current head `b14158884886286f4ed5748b9cd03bd7f228ce47` (no prior review existed at any head).

**Disposition: merge.**

- The functional change is correct and verified against `main`: in `horde_ac` mode the pipeline horde *is* the AC critic (`ac_state.replace(critic_state=state.horde_state)`), so `self._horde.horde_spec.gammas[value_head_index]` is exactly the value head's own gamma. `HordeActorCriticAgent.update(discount=...)` on main already implements the termination-aware machinery (`carry_traces = value_discount != 0.0`, zeroed bootstrap at discount 0) — the pipeline simply never forwarded `terminated`, so `horde_ac` mode bootstrapped value and carried actor traces across episode boundaries. Off-terminal steps forward the critic's own gamma (semantics unchanged); terminal steps now pass discount 0.
- The new end-to-end test asserts both properties: terminal value-head TD target equals the reward alone (0.5, exact) with terminal actor traces zeroed, and the non-terminal target differs.
- CI at the head is 30/30 green (runtime shards on both Python versions + ci-ok). `pipeline.py` is not in the evidence-registry source inventory, so no pinned-artifact implications.

One non-blocking note: roughly 85 of the 100 added lines are pure reformatting of untouched code (collapsing multi-line expressions). It passes the 100-column lint, but please keep future diffs limited to the changed logic — wide churn makes exact-head review harder in a repository that binds evidence to source identity.

**Review by lumix5 — COMMENTED** [review]
**Reconciliation note from the same contributor** — two concurrent review sessions posted under this account ~18 minutes apart (the earlier COMMENTED disposition and the later approve). Consolidated position for maintainers: the *technical* assessment in the later review stands (the fix itself is verified correct); the *process* point in the earlier one also stands — this PR duplicates a first-created, already-verified twin. #2528 is the first-created identical fix and already carries an independent 184-test verification; consolidating on #2528 and closing this one loses nothing. If maintainers prefer this copy, #2528 should close instead — one of the two must.

**Comment by lalalune** [comment]
Reviewed: Superseded by #2348 after diff review: same terminal value-head discount behavior without unrelated whole-file formatting churn. Retained implementation passes all 184 pipeline tests locally.
